File: src/lessons/welcome/array/basics2/Array2Entity.py
def setX(i):
	errorMsg("Sorry Dave, I cannot let you run setX(i) in this exercise")

def setY(i):
	errorMsg("Sorry Dave, I cannot let you run setY(i) in this exercise")
def setPos(x,y):
	errorMsg("Sorry Dave, I cannot let you run setPos(x,y) in this exercise")

# BEGIN TEMPLATE
# BEGIN SOLUTION
def mark(color):
	setBrushColor(color)
	brushDown()
	brushUp()

colors = []

# read the colors
for i in range(getWorldHeight()-1):
	colors.append(getGroundColor())
	forward()
colors.append(getGroundColor())
backward(getWorldHeight()-1)

def makeLine (colors):
	offset = int(readMessage())
	mark(colors[(0+offset)% colors.__len__()])
	for i in range(getWorldWidth()-1):
		forward()
		mark(colors[(i+1+offset)% colors.__len__()])
	backward(getWorldWidth()-1)

# duplicate the pattern
for i in range(getWorldWidth()-1):
	turnLeft()
	forward()
	turnRight()
	makeLine(colors)

# END SOLUTION
# END TEMPLATE

File: src/lessons/welcome/array/basics/Array2Entity.scala
package lessons.welcome.array.basics;

import java.awt.Color
import plm.universe.bugglequest.SimpleBuggle;
import plm.core.model.Game

class ScalaArray2Entity extends SimpleBuggle {
	override def setX(i: Int)  {
		if (isInited)
			throw new RuntimeException(Game.i18n.tr("I'm sorry Dave, I'm affraid I can't let you use setX() in this exercise."));
	}
	override def setY(i: Int)  { 
		if (isInited)
			throw new RuntimeException(Game.i18n.tr("I'm sorry Dave, I'm affraid I can't let you use setY() in this exercise."));
	}
	override def setPos(x: Int, y:Int)  { 
		if (isInited)
			throw new RuntimeException(Game.i18n.tr("I'm sorry Dave, I'm affraid I can't let you use setPos() in this exercise."));
	}

	/* BEGIN TEMPLATE */
	def run() {
	/* BEGIN SOLUTION */
		val colors = new Array[Color](getWorldHeight());

		/* read the colors */
		colors(0) = getGroundColor();
		for (i <- 1 to getWorldHeight()-1) {
			forward();
			colors(i) = getGroundColor();
		}
		backward(getWorldHeight()-1);

		/* duplicate the pattern */
		for (col <- 1 to getWorldWidth()-1) {
			left();
			forward();
			right();
			makeLine(colors);
		}
	}

	def makeLine(colors: Array[Color]) {
		val offset = readMessage().toInt;
		mark(colors( (0+offset)%colors.length ) );
		for (i <- 1 to getWorldWidth()-1) {
			forward();
			mark(colors(  (i+offset)%colors.length  ));
		}
		backward(getWorldHeight()-1);
	}
	def mark(c:Color){
		setBrushColor(c);
		brushDown();
		brushUp();
		/* END SOLUTION */
	}
	/* END TEMPLATE */
}
